extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/RouteHandler.java

```diff
@@ -33,13 +34,23 @@ default void handle(RoutingContext context) {
             invoke(context);
         } else {
             try {
+                // Activate the context, i.e. set the thread locals
                 requestContext.activate();
                 if (user != null) {
                     RouteHandlers.fireSecurityIdentity(user.getSecurityIdentity());
                 }
+                // Reactive routes can use async processing (e.g. mutiny Uni/Multi) and context propagation
+                // 1. Store the state (which is basically a shared Map instance)
+                // 2. Terminate the context correcly when the response is disposed or an exception is thrown 
+                InjectableContext.ContextState state = requestContext.getState();
+                context.addEndHandler(v -> {
+                    requestContext.activate(state);
```

Comment:
Wouldn't it be more performant to have a terminate(state) method in arc, so you don't have to touch the thread locals?

Reply:
Also we try and avoid lambdas in runtime code.

Reply:
> so you don't have to touch the thread locals?

Hm, yes that would help. We don't need to activate the context here.